src/flint/types/fmpz_mod.pyx

```diff
@@ -0,0 +1,428 @@
+            if not self.ctx.val.n == (<fmpz_mod>other).ctx.val.n:
+                raise ValueError("moduli must match")
+
+            fmpz_mod_add(
+                self.val, self.val, (<fmpz_mod>other).val, self.ctx.val
+            )
+            return self
+
+        # Add an fmpz to an fmpz_mod
+        other = any_as_fmpz(other)
+        if other is NotImplemented:
+            raise NotImplementedError
+
+        fmpz_mod_add_fmpz(
+            self.val, self.val, (<fmpz>other).val, self.ctx.val
+        )
+        return self
+
+    def __sub__(self, other):
+        cdef fmpz_mod res
+        res = fmpz_mod.__new__(fmpz_mod)
+        res.ctx = self.ctx
+
+        # Sub two fmpz_mod if moduli match
+        if typecheck(other, fmpz_mod):
+            if not self.ctx.val.n == (<fmpz_mod>other).ctx.val.n:
+                raise ValueError("moduli must match")
+
+            fmpz_mod_sub(
+                res.val, self.val, (<fmpz_mod>other).val, self.ctx.val
+            )
+            return res
+
+        # Attempt to sub an fmpz to an fmpz_mod
+        other = any_as_fmpz(other)
+        if other is NotImplemented:
+            raise NotImplementedError
+
+        fmpz_mod_fmpz_sub(
+            res.val, self.val, (<fmpz>other).val, self.ctx.val
+        )
+
+        return res
+
+    # TODO: is this bad? Should I just copy paste logic
+    # above?
+    def __rsub__(self, other):
+        return self.__sub__(other).__neg__()
+
+    def __isub__(self, other):
+        if typecheck(other, fmpz_mod):
+            if not self.ctx.val.n == (<fmpz_mod>other).ctx.val.n:
+                raise ValueError("moduli must match")
+
+            fmpz_mod_sub(
+                self.val, self.val, (<fmpz_mod>other).val, self.ctx.val
+            )
+            return self
+
+        other = any_as_fmpz(other)
+        if other is NotImplemented:
+            raise NotImplementedError
+
+        fmpz_mod_sub_fmpz(
+            self.val, self.val, (<fmpz>other).val, self.ctx.val
+        )
+        return self
+
+    def __mul__(self, other):
+        cdef fmpz_mod res
+        res = fmpz_mod.__new__(fmpz_mod)
+        res.ctx = self.ctx
+
+        # Add two fmpz_mod if moduli match
+        if typecheck(other, fmpz_mod):
+            if not self.ctx.val.n == (<fmpz_mod>other).ctx.val.n:
+                raise ValueError("moduli must match")
+
+            fmpz_mod_mul(
+                res.val, self.val, (<fmpz_mod>other).val, self.ctx.val
+            )
+            return res
+
+        # Attempt to add an fmpz to an fmpz_mod
+        other = any_as_fmpz(other)
+        if other is NotImplemented:
+            raise NotImplementedError
+
+        # There's no `fmpz_mod_mul_fmpz`
+        # So we need to make sure that other is canonical
+        # TODO: should we check with `fmpz_mod_is_canoncial`
+        # or just always reduce?
+        fmpz_mod_set_fmpz(
+            (<fmpz>other).val, (<fmpz>other).val, self.ctx.val
+        )
```

Comment:
I think that there should be a function like `any_as_fmpz_mod` or perhaps a `cdef` method like `self.any_as_fmpz_mod(other)`.

I wouldn't worry too much about the performance of `a * b` for mixed types. If someone wants maximum speed they should be ensuring that `a` and `b` are already the same type, same modulus etc.

Reply:
OK, I'm glad you said this as it was one of my questions. If I implement this then the arithmetic simplifies nicely as I'll do the conversion to `fmpz_mod` and then have the simplier cases.

Reply:
In fact, the "Easiest" method is simply to call `other = self.ctx(other)` which does all the type conversions. 

A faster(?) method would be a `cdef` method which does things with the `cython` rather than `python` types I suppose.